controller/src/main/java/org/jboss/as/controller/access/constraint/PatternScopedConstraint.java

```diff
@@ -0,0 +1,215 @@
+
+    /** Constructor for the global user constraint */
+    private PatternScopedConstraint() {
+        super();
+        this.user = true;
+        this.global = true;
+        this.readOnly = false;
+        this.readOnlyConstraint = null;
+        this.patternsHolder = new PatternsHolder();
+        this.target = null;
+    }
+
+    /** Constructor for the required constraint */
+    private PatternScopedConstraint(String target) {
+        super();
+        this.user = false;
+        this.global = false;
+        this.patternsHolder = null;
+        this.target = target;
+        this.readOnly = false;
+        this.readOnlyConstraint = null;
+    }
+
+    /** Constructor for the management layer to use when new roles are created. */
+    public PatternScopedConstraint(List<Pattern> allowed) {
+        super();
+        this.user = true;
+        this.global = false;
+        this.patternsHolder = new PatternsHolder(allowed);
+        this.target = null;
+        this.readOnly = false;
+        this.readOnlyConstraint = new PatternScopedConstraint(this.patternsHolder, true);
+    }
+
+    /**
+     * Creates the constraint the standard constraint will return from {@link #getOutofScopeReadConstraint()}
+     * Only call from {@link PatternScopedConstraint#PatternScopedConstraint(java.util.List)}
+     */
+    private PatternScopedConstraint(PatternsHolder patternsHolder, boolean readOnly) {
+        super();
+        this.user = true;
+        this.global = false;
+        this.patternsHolder = patternsHolder;
+        this.target = null;
+        this.readOnly = readOnly;
+        this.readOnlyConstraint = null;
+    }
+
+    public void setAllowedPatterns(List<Pattern> allowed) {
+        assert !global : "constraint is global";
+        assert readOnlyConstraint != null : "invalid cast";
+        this.patternsHolder.patterns = new LinkedHashSet<>(allowed);
+    }
+
+    @Override
+    public boolean violates(Constraint other, Action.ActionEffect actionEffect) {
+        if (other instanceof PatternScopedConstraint) {
+            PatternScopedConstraint psc = (PatternScopedConstraint) other;
+            if (user) {
+                assert !psc.user : "illegal comparison";
+                if (!readOnly) {
+                    if (!global) {
+                        if (actionEffect == Action.ActionEffect.WRITE_RUNTIME || actionEffect == Action.ActionEffect.WRITE_CONFIG) {
+                            Set<Pattern> ourPatterns = patternsHolder.patterns;
+                            String theTarget = psc.target;
+                            //  Ok as long as one of our patterns match
+                            boolean anyMatch = anyMatch(ourPatterns, theTarget);
+                            if (!anyMatch) ControllerLogger.ACCESS_LOGGER.tracef("pattern constraint violated " +
+                                            "for action %s due to no match between target %s and allowed patterns %s",
```

Comment:
Could that possibly leak information, if the patterns target non-addressable resources? That should be unlikely, but the possibility occured to me immediately :-)